beacon/src/region.rs

```diff
@@ -0,0 +1,203 @@
+            where
+                E: de::Error,
+            {
+                let proto_region = ProtoRegion::from_str(value)
+                    .map_err(|_| de::Error::custom(format!("unsupported region: {value}")))?;
+                Ok(Region(proto_region))
+            }
+        }
+
+        deserializer.deserialize_str(RegionVisitor)
+    }
+}
+
+impl fmt::Display for Region {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<Region> for i32 {
+    fn from(region: Region) -> Self {
+        region.0.into()
+    }
+}
+
+impl From<&Region> for i32 {
+    fn from(region: &Region) -> Self {
+        region.0.into()
+    }
+}
+
+impl Region {
+    pub fn from_i32(v: i32) -> Result<Self> {
+        ProtoRegion::from_i32(v)
+            .map(Self)
+            .ok_or_else(|| Error::unsupported_region(v))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RegionParams {
+    pub gain: Decimal,
+    pub region: Region,
+    pub params: Vec<BlockchainRegionParamV1>,
+}
+
+impl AsRef<[BlockchainRegionParamV1]> for RegionParams {
+    fn as_ref(&self) -> &[BlockchainRegionParamV1] {
+        self.params.as_ref()
+    }
+}
+
+impl PartialEq for RegionParams {
+    fn eq(&self, other: &Self) -> bool {
+        self.gain.eq(&other.gain) && self.region.eq(&other.region) && self.params.eq(&other.params)
+    }
+}
+
+impl TryFrom<GatewayRegionParamsStreamedRespV1> for RegionParams {
+    type Error = Error;
+    fn try_from(value: GatewayRegionParamsStreamedRespV1) -> Result<Self> {
+        let region = Region::from_i32(value.region)?;
+        let params = value
+            .params
+            .ok_or_else(Error::no_region_params)?
+            .region_params;
+        Ok(Self {
+            gain: Decimal::new(value.gain as i64, 1),
+            params,
+            region,
+        })
+    }
+}
+
+impl TryFrom<GatewayRegionParamsRespV1> for RegionParams {
+    type Error = Error;
+    fn try_from(value: GatewayRegionParamsRespV1) -> Result<Self> {
+        let region = Region::from_i32(value.region)?;
+        let params = value
+            .params
+            .ok_or_else(Error::no_region_params)?
+            .region_params;
+        Ok(Self {
+            gain: Decimal::new(value.gain as i64, 1),
+            params,
+            region,
+        })
+    }
+}
+// This is the currently minimimum conducted power supported by the semtech
+// packet forwarder
+lazy_static::lazy_static! {
+    static ref MIN_CONDUCTED_POWER: Decimal = Decimal::new(120, 1);
+}
+
+impl RegionParams {
+    pub fn max_eirp(&self) -> Result<Decimal> {
+        self.params
+            .iter()
+            .max_by_key(|p| p.max_eirp)
+            .ok_or_else(Error::no_region_params)
+            .map(|v| Decimal::new(v.max_eirp as i64, 1))
+    }
+
+    pub fn max_conducted_power(&self) -> Result<u32> {
+        self.max_eirp().and_then(|max_eirp| {
+            (max_eirp - self.gain)
```

Comment:
Seeing this helps me feel more relieved about my earlier comment, but what is self.gain in this function? Is that the local gateway's asserted antenna gain?

Reply:
yes that's the hotspots's asserted antenna gain. When gateway-rs gets "region_params" for it's public key it gets back the asserted region, gain and region_params for the region